hbase_storage/src/main/java/com/splicemachine/si/impl/server/SICompactionStateMutate.java

```diff
@@ -116,6 +120,27 @@ private void mutate(Cell element, TxnView txn) throws IOException {
                     assert deleteRightAfterFirstWriteTimestamp == 0;
                     deleteRightAfterFirstWriteTimestamp = element.getTimestamp();
                     break;
+                case USER_DATA: {
+                    long t = element.getTimestamp();
```

Comment:
Here we are considering the begin timestamp of the transaction that wrote the data, but we need to consider the commit timestamp instead (there's the same problem with deletes too, https://splicemachine.atlassian.net/browse/DB-9640 )

Consider the lowWaterMark is 1000 (and there is a transaction with timestamp 1000) 
If we are processing an update with timestamp = 900 and commit timestamp 1200, we are adding it to the map. Then we remove the baseline insert @ ts 100. 

When the lowWatermark transaction @ 1000 goes to read, there's no baseline anymore and the row is gone.

I think the race condition is pretty hard to hit but better to get it fixed.

Reply:
Very good point. 
I created separate PRs for DB-9640 against 2.8/3.0/master that address the deleted rows issue.
I also updated this (DB-8781) to include the fix and extend it to updates.